File: 剑指OfferII053二叉搜索树中的中序后继.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderSuccessor(self, root: 'TreeNode', p: 'TreeNode') -> 'TreeNode':
        res = None

        while root:
            if root.val > p.val:    # 大走左节点，因为二叉搜索树左子树都小于根节点
                res = root
                root = root.left    # 直到找到小于对应根节点的值中最小的
            else:                   # 否则走右子树
                root = root.right
        return res

File: 剑指OfferII053二叉搜索树中的中序后继.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderSuccessor(root *TreeNode, p *TreeNode) *TreeNode {
	res := &TreeNode{}
	res = nil // 需要防止不存在的情况，该情况下 Val 会有默认值 0
	for root != nil {
		if root.Val > p.Val {
			res = root
			root = root.Left
		} else {
			root = root.Right
		}
	}
	return res
}